### PR TITLE
Archery adapter

### PR DESCRIPTION
**Pull request by alistaire47**
This moves the `cpp-micro` benchmark to use [ArcheryAdapter](https://github.com/conbench/conbench/blob/main/benchadapt/python/benchadapt/adapters/archery.py), so it will handle all its own measuring, munging, and posting to conbench.

As of writing, my arrow install is still breaking such that I can't call `archery benchmark run` properly, so I can't do a full `conbench cpp-micro` end-to-end test, which we should presumably find a way to do before merging.

### CODE REVIEW

**Comment by alistaire47** [comment]
Ah crap, I realized I think I forgot to copy all the `arrow_info` type metadata over to `ArcheryAdapter`...I'll probably need a little PR to fix that

**Comment by jonkeane** [comment]
This is fantastic! Thank you. Especially love:
image

We should wait on merging this until the `arrow_info` is copied over to `ArcheryAdapter`, yeah?

Then we can run a test benchmark at https://buildkite.com/apache-arrow/arrow-bci-benchmark-build-test yeah?

**Comment by alistaire47** [comment]
> We should wait on merging this until the `arrow_info` is copied over to `ArcheryAdapter`, yeah?

Hmm, so currently [arrow_info()](https://github.com/voltrondata-labs/benchmarks/blob/main/benchmarks/_benchmark.py#L30-L47) uses `pyarrow.cpp_build_info`, and I'm not sure 

1. if the pyarrow version will necessarily correspond to the C++ code being benchmarked (at certain points with my installs it certainly wouldn't; on CI I'm less worried), and
2. if this should actually be part of `ArcheryAdapter`, since 
    a. it would add a dep on pyarrow we don't want, and
    b. it would make it explicitly arrow-centric, which brings into question why it lives in `benchadapt` at all.

So probably we should insert build info from here, but I'm still not sure where the best place to get it from is. We could keep getting it from pyarrow, I guess, which would be the same as what we're currently doing.

**Comment by jonkeane** [comment]
This seems like another place where a user of an adapter might need to craft their own version setters, yeah?

**Comment by alistaire47** [comment]
[This passed finally!](https://buildkite.com/apache-arrow/arrow-bci-benchmark-on-arm64-m6g-linux-compute/builds/1402#01838f1c-9a34-4446-87fa-f2dcc6f504af/6-13905) So it's ready for review again. I wouldn't necessarily use this as an example of how adapters should be called; there's a lot of mess here to make it integrate with {benchmarks}.

File: benchmarks/cpp_micro_benchmarks.py [hunk 1]
```diff
@@ -114,47 +88,23 @@ class RecordCppMicroBenchmarks(_benchmark.Benchmark):
...
-            info,
-            context,
-            options=options,
-            output=result,
-            name=name,
+    def __init__(self):
+        tags, info, context = self._get_tags_info_context(case=None, extra_tags={})
+
+        self.adapter = ArcheryAdapter(
+            result_fields_override={"github": self.github_info},
```
Comment (jonkeane) [inline benchmarks/cpp_micro_benchmarks.py:97]: We might want to have a comment here talking about why we're relying on github_info from _benchmark.py

File: benchmarks/cpp_micro_benchmarks.py [hunk 2]
```diff
@@ -114,47 +88,23 @@ class RecordCppMicroBenchmarks(_benchmark.Benchmark):
...
-        return self.record(
-            values,
-            tags,
-            info,
-            context,
-            options=options,
-            output=result,
-            name=name,
+    def __init__(self):
+        tags, info, context = self._get_tags_info_context(case=None, extra_tags={})
```
Comment (jonkeane) [inline benchmarks/cpp_micro_benchmarks.py:94]: Same thing as below, a comment about where | why we are depending on this here might be nice

**Review by jonkeane — COMMENTED** [review]
Looking good — even with integrating with the current setup we're at +61, -157 (and much of that +61 is boiler plate packaging stuff 🎉

File: benchmarks/cpp_micro_benchmarks.py (hunk 2 above)
Comment (alistaire47) [reply benchmarks/cpp_micro_benchmarks.py:94]: Done. This appends a bunch of metadata about arrow like build flags, all extracted from pyarrow. I don't love that for a C++ benchmark, but a. it's consistent, and b. I'm not sure how else to dig it out.

Comment (jonkeane) [reply benchmarks/cpp_micro_benchmarks.py:95]: _nods_ yeah it's a bit odd, but in the end probably the easiest. We run into a similar issue with velox, where there's not (yet) an easy thing to poll for these kinds of things. But they are so project-specific it's not something we can really abstract.

**Review by jonkeane — APPROVED** [review]

**Pull request merged**